onnxruntime/core/providers/shared_library/provider_interfaces.h

```diff
@@ -0,0 +1,480 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// Public wrappers around internal ort interfaces (currently)
+// In the future the internal implementations could derive from these to remove the need for the wrapper implementations
+
+#include "core/framework/func_api.h"
+
+namespace ONNX_NAMESPACE {
+enum AttributeProto_AttributeType : int;
+enum OperatorStatus : int;
+
+// String pointer as unique TypeProto identifier.
+using DataType = const std::string*;
+
+struct Provider_TensorProto {
+  virtual ~Provider_TensorProto() = default;
+
+  virtual void CopyFrom(const Provider_TensorProto& v) = 0;
+
+  void operator=(const Provider_TensorProto& v) { CopyFrom(v); }
+};
+
+struct Provider_AttributeProto {
+  static std::unique_ptr<Provider_AttributeProto> Create();
+
+  virtual ~Provider_AttributeProto() = default;
+  virtual std::unique_ptr<Provider_AttributeProto> Clone() const = 0;
+
+  virtual ::onnx::AttributeProto_AttributeType type() const = 0;
+  virtual int ints_size() const = 0;
+  virtual int64_t ints(int i) const = 0;
+  virtual int64_t i() const = 0;
+  virtual float f() const = 0;
+  virtual void set_s(const ::std::string& value) = 0;
+  virtual const ::std::string& s() const = 0;
+  virtual void set_name(const ::std::string& value) = 0;
+  virtual void set_type(::onnx::AttributeProto_AttributeType value) = 0;
+  virtual ::onnx::Provider_TensorProto* add_tensors() = 0;
+
+  void operator=(const Provider_AttributeProto& v) = delete;
+};
+
+// This is needed since Provider_NodeAttributes is a map of unique_ptr to Provider_AttributeProto and that won't work since unique_ptrs are not copyable
+// (supposedly this should work in the latest C++ STL but it didn't for me so I used this to make it copyable)
+struct Provider_AttributeProto_Copyable {
+  Provider_AttributeProto_Copyable() = default;
+  Provider_AttributeProto_Copyable(const Provider_AttributeProto_Copyable& copy) : p_{copy->Clone()} {}
+
+  void operator=(std::unique_ptr<Provider_AttributeProto>&& p) { p_ = std::move(p); }
+  void operator=(const Provider_AttributeProto_Copyable& p) { p_ = p->Clone(); }
+
+  Provider_AttributeProto& operator*() const { return *p_.get(); }
+  Provider_AttributeProto* operator->() const { return p_.get(); }
+
```

Comment:
Should these check there's a value in p_?

Reply:
It'll be nullptr if it's not set. It's an internal thing I want to remove in the future so I don't think it'd really be worth it.